packages/query-store/src/QueryStore.ts

```diff
@@ -0,0 +1,340 @@
+	 * Therefore, there is a tick when the array is full of undefined, because we are still fetching.
+	 * In that time period, we return this empty object so we don't get `cannot access x of undefined` errors.
+	 * This may break in instances where the user may have nested arrays / dicts
+	 */
+	#mockResult: QueryResult = {};
+
+	readonly id: string;
+
+	constructor(
+		query: string | Query,
+		exec: Runner,
+		id?: string,
+		private readonly opts: QueryStoreOpts = { disableCache: false },
+		private readonly root?: QueryStore
+	) {
+		super();
+		Object.freeze(opts);
+		// Ensure an ID Exists
+		if (!id) this.id = buildId(query);
+		else this.id = id;
+
+		// TODO: Strip any trailing ; from queries
+		// This is hard because of comments
+		// We might want to just error out if the querystring contains a ; for simplicity
+
+		if (typeof query === 'string') {
+			this.#query.from({ __userQuery: sql`(${query})` }).select('*');
+		} else this.#query = query;
+		// @ts-expect-error Passing undocumented parameter
+		this.#exec = (...args: Parameters<Runner>) => exec(args[0], args[1] ?? this.id);
+
+		this.#proxied = new Proxy<QueryStore & QueryResult[]>(
+			this as unknown as QueryStore & QueryResult[],
+			{
+				get: (self, _prop) => {
+					// Intercept numeric indices. This implies we're trying to access rows (data) in the store.
+					// If the data has not been loaded, initiate the async #update method to fetch the data.
+					let prop: string | symbol | number = _prop;
+					if (typeof prop === 'string' && /^[\d.]+$/.exec(prop)) prop = parseInt(prop);
+					if (typeof prop === 'number') {
+						if (!self.#loaded) {
+							try {
+								const r = self.#fetchData();
+								if (r instanceof Promise)
+									r.catch((e) => {
+										throw new Error('Failed to update query store', { cause: e });
+									});
+							} catch (e) {
+								throw new Error('Failed to update query store', { cause: e });
+							}
+						}
+
+						if (!self.#values[prop]) return self.#mockResult;
+						return self.#values[prop];
+					}
+
+					// Intercept 'length' property. This implies we're trying to get the total number of rows (data) in the store.
+					// If the length has not been correctly updated, initiate the async #updateLength method to calculate it.
+					else if (prop === 'length') {
+						// TODO: Conditional
+						if (!self.#lengthLoaded) {
+							try {
+								const r = self.#fetchLength();
+								if (r instanceof Promise)
+									r.catch((e) => {
+										throw new Error('Failed to update query store length', { cause: e });
+									});
+							} catch (e) {
+								throw new Error('Failed to update query store length', { cause: e });
+							}
+						}
+					}
+
+					if (prop in self) {
+						// @ts-expect-error Typescript gets mad about this for some reason
+						if (typeof self[prop] === 'function') return self[prop].bind(self);
```

Comment:
Maybe `ts` wants you to check that `self[prop` is not `undefined`

Reply:
The type error that was was that `prop` may not be a string, and typescript expects `self` to only be index accessed by numbers